{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def gene_annotate(filename, ref_file):\n",
    "    '''\n",
    "    This function will create a dataframe from a vcf file and add a column with the genes associated with that variant.\n",
    "    \n",
    "\n",
    "    This function takes a filename and reference file name location as a string. The filename corresponds to gene variations\n",
    "    in a cancer subtype and the reference file name corresponds to genes and their positions in the human genome. This\n",
    "    function will read both files into the program as dataframes, eliminating irrelevant data. The function will then\n",
    "    iterate through the filename dataframe to determine genes associated with each variant (referencing the gene information\n",
    "    in the ref_file dataframe) and attach the genes associated as a new column to the dataframe.\n",
    "    \n",
    "    Parameters: \n",
    "    \n",
    "    ref_file (string): This argument is a (.vcf) file location as a string corresponding to gene variations present\n",
    "                       in a cancer subset.\n",
    "    \n",
    "    \n",
    "    filename (string): This argument is (.txt) file location as a string corresponding to genes in the\n",
    "                       human genome.\n",
    "\n",
    "\n",
    "    Returns: \n",
    "    data_frame (DataFrame): The output is a DataFrame consisting of each variant from the original (.vcf) file with the\n",
    "                            'CHROM', 'POS', 'REF', 'ALT' columns, as well as, a new 'GENE_SYMBOLS' column corresponding\n",
    "                            to the genes asssociated with that genetic variant.\n",
    "                            \n",
    "   \n",
    "    \n",
    "    '''\n",
    "    data_frame=pd.read_csv(filename, header=None, sep='\\t',index_col=None, comment='#', names=['CHROM','POS','ID','REF','ALT','QUAL','FILTER','INFO','FORMAT','GROUP_C'],usecols=['CHROM','POS','REF','ALT'])\n",
    "    genesymbol_df = pd.read_csv(ref_file, header=0, sep='\\t',index_col=None)\n",
    "    master_gene_list = []\n",
    "    for index, row in data_frame.iterrows(): \n",
    "        position = row [\"POS\"]\n",
    "        gene_df_subset = genesymbol_df.loc[(genesymbol_df['START'] <= position) & (genesymbol_df['STOP'] >= position)]\n",
    "        row_genes = gene_df_subset[\"GENESYMBOL\"].tolist()\n",
    "        row_genes = list(set(row_genes))\n",
    "        master_gene_list.append(row_genes)\n",
    "    data_frame['GENE_SYMBOL'] = master_gene_list\n",
    "    data_frame['GENE_SYMBOL'] = data_frame['GENE_SYMBOL'].apply(lambda x: np.nan if len(x) == 0 else tuple(x))\n",
    "    return data_frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def common_variants(dataframe1, dataframe2, dataframe3):\n",
    "    \n",
    "    '''\n",
    "    This function will create a dataframe of gene variants from three separate dataframes (created from (.vcf) files of a \n",
    "    specific cancer subtype) for variants that are present in all three cancer subtypes.\n",
    "    \n",
    "\n",
    "    This function takes three dataframes, each corresponding to a (.vcf) file for gene variants in a specific cancer subtype,\n",
    "    and first merges the first and second dataframes, then the merged dataframe is merged with the third dataframe. Since\n",
    "    pd.merge is used with how='inner', when this occurs only the intersection of the dataframes is kept in the new dataframe,\n",
    "    therefore, returning only the common variants among the three cancer subtypes\n",
    "    Parameters: \n",
    "    \n",
    "    dataframe1 (DataFrame): This argument is a DataFrame, corresponding to a (.vcf) file of gene variants present\n",
    "                            in a cancer subset.\n",
    "    \n",
    "    \n",
    "    dataframe2 (DataFrame): This argument is a DataFrame, corresponding to a (.vcf) file of gene variants present\n",
    "                            in a cancer subset.\n",
    "                            \n",
    "    dataframe3 (DataFrame): This argument is a DataFrame, corresponding to a (.vcf) file of gene variants present\n",
    "                            in a cancer subset.\n",
    "\n",
    "\n",
    "    Returns: \n",
    "    variants (DataFrame): The output is a DataFrame consisting of every gene variant common among the three input cancer\n",
    "                          subsets.\n",
    "                            \n",
    "   \n",
    "    \n",
    "    '''\n",
    "    variants1=pd.merge(dataframe1, dataframe2, on=None, how='inner')\n",
    "    variants=pd.merge(variants1,dataframe3, on=None, how='inner')\n",
    "    return variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def specific_variants(dataframe1, dataframe2, dataframe3):\n",
    "    \n",
    "    '''\n",
    "    This function will create a dataframe of gene variants from three separate dataframes (created from (.vcf) files of a \n",
    "    specific cancer subtype) for variants that are ONLY present in the cancer subtype corresponding to the first Dataframe,\n",
    "    dataframe1, and NOT present in the second or third cancer subset.\n",
    "    \n",
    "\n",
    "    This function takes three dataframes, each corresponding to a (.vcf) file for gene variants in a specific cancer subtype,\n",
    "    and first merges the first and second dataframes, then merges the first and third dataframes. Since pd.merge is used with\n",
    "    how='inner', when this occurs only the intersection of the dataframes is kept in the new dataframe, therefore, returning\n",
    "    only the common variants among the two cancer subtypes merged. This is done to determine the commonalities between the\n",
    "    first cancer subtype (the one of interest) and the other two cancer subtypes. The two merged dataframes (looking for \n",
    "    commonalities between the first and second and first and third cancer subtypes) is then concatenated, and\n",
    "    pd.Dataframe.drop_duplicates is used to eliminate any variants that are present in all three dataframes, leaving only the\n",
    "    gene variants related to the first cancer subset (dataframe1).\n",
    "    Parameters: \n",
    "    \n",
    "    dataframe1 (DataFrame): This argument is a DataFrame, corresponding to a (.vcf) file of gene variants present\n",
    "                            in a cancer subset. This should be the cancer subset of which you want to determine\n",
    "                            gene variants related to this cancer subtype ONLY and NOT the other two cancer subtypes.\n",
    "    \n",
    "    \n",
    "    dataframe2 (DataFrame): This argument is a DataFrame, corresponding to a (.vcf) file of gene variants present\n",
    "                            in a cancer subtype.\n",
    "                            \n",
    "    dataframe3 (DataFrame): This argument is a DataFrame, corresponding to a (.vcf) file of gene variants present\n",
    "                            in a cancer subtype.\n",
    "\n",
    "\n",
    "    Returns: \n",
    "    variants (DataFrame): The output is a DataFrame consisting of every gene variant specific to ONLY the first input\n",
    "                          cancer subtype (dataframe1).\n",
    "                            \n",
    "   \n",
    "    \n",
    "    '''\n",
    "    common1 =pd.merge(dataframe1,dataframe2, on=None, how='inner')\n",
    "    common2= pd.merge(dataframe1,dataframe3, on=None, how='inner')\n",
    "    specific_variants=pd.concat([dataframe1,common1,common2]).drop_duplicates(keep=False, inplace=False)\n",
    "    return specific_variants"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***THIS CELL WILL PRINT OUT THE VARIANTS COMMON AMONG ALL THREE CANCER SUBTYPES AS variants_popular, AND THE VARIANTS SPECIFIC TO ONLY TRIPLE NEGATIVE BREAST CANCER***"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    CHROM       POS REF ALT  GENE_SYMBOL\n",
      "0       1    565286   C   T          NaN\n",
      "1       1    879676   G   A    (SAMD11,)\n",
      "2       1    879687   T   C    (SAMD11,)\n",
      "3       1    881627   G   A          NaN\n",
      "4       1    887801   A   G          NaN\n",
      "..    ...       ...  ..  ..          ...\n",
      "64      1  12072518   G   C      (MFN2,)\n",
      "65      1  12267292   C   T  (TNFRSF1B,)\n",
      "66      1  12401868   A   C    (VPS13D,)\n",
      "67      1  13943724   T   C      (PDPN,)\n",
      "68      1  14116491   C   G     (PRDM2,)\n",
      "\n",
      "[69 rows x 5 columns]      CHROM       POS    REF  ALT GENE_SYMBOL\n",
      "0        1    237763      G    A         NaN\n",
      "4        1    567242      G    A         NaN\n",
      "5        1    723801  AGAGA  AGA         NaN\n",
      "6        1    729679      C    G         NaN\n",
      "7        1    751832      T    C         NaN\n",
      "..     ...       ...    ...  ...         ...\n",
      "967      1  14458859      C    T         NaN\n",
      "968      1  14468557      C    T         NaN\n",
      "969      1  14474265      T    A         NaN\n",
      "970      1  14490361      A    G         NaN\n",
      "971      1  14495694      C    G         NaN\n",
      "\n",
      "[751 rows x 5 columns]\n"
     ]
    }
   ],
   "source": [
    "TNBC = gene_annotate('TNBC.vcf','hg19_genes_chr1p_bed.txt')\n",
    "nTNBC = gene_annotate('nTNBC.vcf','hg19_genes_chr1p_bed.txt')\n",
    "HER2 = gene_annotate('HER2.vcf','hg19_genes_chr1p_bed.txt')\n",
    "\n",
    "variants_popular = common_variants(TNBC,nTNBC,HER2)\n",
    "\n",
    "CancerSubtype_ofInterest = specific_variants(TNBC,nTNBC,HER2)\n",
    "\n",
    "print(variants_popular, CancerSubtype_ofInterest)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
