{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Homework 5 - BIOINF 575\n",
    "\n",
    "Goals:\n",
    "* Reading from a file\n",
    "* Creating and using basic data structures - list(), dict()\n",
    "* Creating and using numpy arrays\n",
    "\n",
    "The data file GSE53697_RNAseq_AD.txt is a tab-separated file that contains raw read counts and gene expression values in the form of RPKM values from RNA sequencing of 8 control and 9 Alzheimer's disease (AD) samples (human).\n",
    "\n",
    "The first line in the file is the header that describes the table of values:\n",
    "\n",
    "GeneID - Gene ID\n",
    "\n",
    "GeneSymbol - Gene Symbol\n",
    "\n",
    "8 columns: C1-8_raw - raw read count values for the 8 control samples\n",
    "\n",
    "9 columns: A9-17_raw - raw read count values for the 9 Alzheimer's disease samples\n",
    "\n",
    "8 columns: C1-8_rpkm - RPKM values for the 8 control samples\n",
    "\n",
    "9 columns: A9-17_rpkm - RPKM values for the 9 Alzheimer's disease (AD) samples\n",
    "\n",
    "\n",
    "Reads Per Kilobase of transcript per Million mapped reads (RPKM) is a normalized unit of transcript/gene expression.\n",
    "This is used to normalize for sequencing depth and gene length. Different genes might have more reads just due to the fact that they have more bases that the reads can align to and that needs to be accounted for. Also, if one sample (s1) has more reads than another sample (s2) then the overall expression of all genes in s1 would be higher than in s2. Hence, the two samples cannot be compared if a normalization for sequencing depth is not performed.\n",
    "\n",
    "Given this file, select differentially expressed genes between Alzheimer's disease and control.\n",
    "There are many ways to compute differential expression.\n",
    "Here, to go from expression to differential expression, for each gene you have to compute the ratio between two averages: (1) the average expression of the gene in disease and (2) the average expression of the gene in control.\n",
    "\n",
    "Since these averages might be 0, before you compute the ratio add 1 to the numerator and the denominator.\n",
    "Using the log2 on this ratio will allow us to easily see if a gene is under-expressed (negative log value) or over-expressed (positive log values) and also the base 2 of the logarithm allows us to see when the value in AD is two times higher than the one in control by using the threshold 1 for the log2ratio.\n",
    "\n",
    "Select the differentially expressed genes by writing the following functions.\n",
    "\n",
    "1. Create a function data_list that takes as a parameter the file name and creates three lists:\n",
    "    * lst_genes - the list of genes symbols - the genes symbols are on the second column of each line in the file;\n",
    "    * lst_ge_C - each element of this list is a list - the list of control RPKM values for a gene, from each line in the file, split the line by tab and take elements 19 to 27 from the list resulted from split (these are 0-based indices); \n",
    "    * lst_ge_AD - each element of this list is a list - the list of AD RPKM values for a gene, from each line in the file, split the line by tab and take elements starting from 27 from the list resulted from split (this is a 0-based index). \n",
    "\n",
    "\n",
    "Note: The first line in the file is the header line, don't use it in computation.\n",
    "\n",
    "2. Create a function compute_row_average that takes as input a list of lists with numbers stored as text (lst_ge) transforms the lists of lists in a 2-dimensional array and returns a 1-dimensional array containing the mean of each row. \n",
    "\n",
    "Note: Use the function astype() for an array that contains numbers as strings to set the type to float\n",
    "\n",
    "3. Fill in the missing code in the gene_dictionary function that computes the log 2 of the ratio between the AD and control mean values returned by the function compute_row_average, then selects the genes that have a change greater than or equal to 2 fold (absolute value >= 1 for the log2ratio), and finally computes and returns a dictionary with the selected genes and their log2ratio (each selected gene is a key in the dictionary and the corresponding log2ratio is the value). \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import math"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "def data_list(file_name):\n",
    "    \"\"\" \n",
    "    This function initializes three lists of lists.\n",
    " \n",
    "\n",
    "    This function initializes three lists of lists, lst_genes, lst_ge_C, and lst_ge_AD. It takes a file path as a string \n",
    "    in order to locate the data for use. Each List is initalized from another function \n",
    "            lst_genes: from lstGenes\n",
    "            lst_ge_C:  from lstGeC\n",
    "            lst_ge_AD: from lstGeAD\n",
    "\n",
    "\n",
    "\n",
    "    Parameters: \n",
    "    filename (string): The argument of this function is a an RNAseq data (.txt) file location as a string.\n",
    "    \n",
    "\n",
    "    Returns: \n",
    "    lst_genes (list): This output is a list of the gene symbols for every gene present in the data\n",
    "                      (In order by first occurence)\n",
    "    \n",
    "    lst_ge_C (list): This output is a list of all of the control RPKM values for each gene, separated into lists\n",
    "                     themselves\n",
    "    \n",
    "    lst_ge_C (list): This output is a list of all of the disease RPKM values for each gene, separated into lists\n",
    "                     themselves\n",
    "\n",
    "    \"\"\"\n",
    "    lst_genes=lstGenes(file_name)\n",
    "    lst_ge_C=lstGeC(file_name)\n",
    "    lst_ge_AD=lstGeAD(file_name)\n",
    "    \n",
    "    \n",
    "    return lst_genes, lst_ge_C, lst_ge_AD "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "def lstGenes(file_name):\n",
    "    \"\"\" \n",
    "    This function creates a list of the gene symbols correlating to the genes present in the data file in order\n",
    "    of first occurrence.\n",
    "    \n",
    "\n",
    "    This function takes a file location as a string to open and read in the data from said file. The file is then\n",
    "    parsed for the gene symbols of every gene present in the data file and added to the list, symbollist.\n",
    "\n",
    "\n",
    "\n",
    "    Parameters: \n",
    "    filename (string): The argument of this function is a an RNAseq data (.txt) file location as a string.\n",
    "\n",
    "\n",
    "    Returns: \n",
    "    lst_genes (list): This output is a list of the gene symbols for every gene present in the data\n",
    "    \n",
    "\n",
    "    \"\"\"\n",
    "    symbolList=[]\n",
    "    with open(file_name, 'r') as file:\n",
    "        for line in file:\n",
    "            if not line.startswith(\"GeneID\"):\n",
    "                line=line.strip()\n",
    "                line=line.split('\\t')\n",
    "                symbolList.append(line[1])\n",
    "    return(symbolList)        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "def lstGeC(file_name):\n",
    "    \"\"\" \n",
    "    This function creates a list of the control RPKM values for each gene present in the data file. Gene data is \n",
    "    separated by individual lists within the list.\n",
    "\n",
    "\n",
    "    This function takes a file location as a string to locate, open, and read in the data from said file. The file is\n",
    "    then parsed for the control RPKM values for each gene. All values per gene are enclosed in a list, before being added\n",
    "    to the final list, C_RPKM_list, and returned.\n",
    "\n",
    "\n",
    "\n",
    "    Parameters: \n",
    "    filename (string): The argument of this function is a an RNAseq data (.txt) file location as a string.\n",
    "\n",
    "\n",
    "    Returns: \n",
    "    lst_ge_C (list): This output is a list of all of the control RPKM values for each gene, separated into lists\n",
    "                     themselves    \n",
    "\n",
    "    \"\"\"\n",
    "    C_RPKM_list=[]\n",
    "    with open(file_name, 'r') as file:\n",
    "        for line in file:\n",
    "            RPKM_list=[]\n",
    "            if not line.startswith(\"GeneID\"):\n",
    "                line=line.strip()\n",
    "                line=line.split('\\t')\n",
    "                line=np.array(line)\n",
    "                line=line[[19,20,21,22,23,24,25,26]]\n",
    "                line = line[np.newaxis,:]\n",
    "                RPKM_list.append(line)\n",
    "                C_RPKM_list.append(RPKM_list)\n",
    "    return(C_RPKM_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "def lstGeAD(file_name):\n",
    "    \"\"\" \n",
    "    This function creates a list of the disease RPKM values for each gene present in the data file. Gene data is \n",
    "    separated by individual lists within the list.\n",
    "\n",
    "\n",
    "    This function takes a file location as a string to locate, open, and read in the data from said file. The file is\n",
    "    then parsed for the disease RPKM values for each gene. All values per gene are enclosed in a list, before being added\n",
    "    to the final list, AD_RPKM_list, and returned.\n",
    "\n",
    "\n",
    "\n",
    "    Parameters: \n",
    "    filename (string): The argument of this function is a an RNAseq data (.txt) file location as a string.\n",
    "\n",
    "\n",
    "    Returns: \n",
    "    lst_ge_C (list): This output is a list of all of the disease RPKM values for each gene, separated into lists\n",
    "                     themselves    \n",
    "\n",
    "    \"\"\"\n",
    "    AD_RPKM_list=[]\n",
    "    with open(file_name, 'r') as file:\n",
    "        for line in file:\n",
    "            RPKM_list=[]\n",
    "            if not line.startswith(\"GeneID\"):\n",
    "                line=line.strip()\n",
    "                line=line.split('\\t')\n",
    "                line=np.array(line)\n",
    "                line=line[[27,28,29,30,31,32,33,34,35]]\n",
    "                line = line[np.newaxis,:]\n",
    "                RPKM_list.append(line)\n",
    "                AD_RPKM_list.append(RPKM_list)\n",
    "    return(AD_RPKM_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_row_average(RPMKlist):\n",
    "    \"\"\" \n",
    "    This function takes a list of lists consisting of RPKM data as an argument. It outputs the mean of the data\n",
    "    for each gene\n",
    "\n",
    "\n",
    "    This function takes a list of lists consisting of RPKM data as an argument. The list is parsed for the numpy arrays\n",
    "    containing the RPKM data for each gene and concatenates them into a 2-dimensional array by order of occurence(Each \n",
    "    row being the RPKM data for a different gene). The mean of each row is taken (therefore, the mean of the RPKM values\n",
    "    for each gene is taken) and subsequently added to the array, geneRPMKmean, which is then returned.\n",
    "\n",
    "\n",
    "\n",
    "    Parameters: \n",
    "    RPMKlist (list): The argument of this function is a list containing lists of RPKM values for the genes in the\n",
    "                     data file\n",
    "\n",
    "\n",
    "    Returns: \n",
    "    geneRPMKmean (array): This output is an array of the mean RPMK values for the genes in the argument list(In the\n",
    "                          order of first occurence)\n",
    "\n",
    "    \"\"\"\n",
    "    rows=len(RPMKlist)\n",
    "    columns=len(RPMKlist[0][0])\n",
    "    matrix=RPMKlist[0][0]\n",
    "    for element in RPMKlist[1:]:\n",
    "        matrix=np.concatenate((matrix,element[0]), axis=0)\n",
    "    matrix=matrix.astype(np.float64)\n",
    "    geneRPMKmean=matrix.mean(axis=1)\n",
    "    return(geneRPMKmean)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "def gene_dictionary(file_name):\n",
    "    \"\"\" \n",
    "    This function creates a dictionary of genes with a 2-fold or more change in gene expession in a disease state\n",
    "    compared to the control. Negative values indicate genes in which expression is two-fold or greater in the \n",
    "    control compared to the disease state. Positive values indicate genes in which expression is two-fold or\n",
    "    greater in the disease state compared to the control.\n",
    "    \n",
    "\n",
    "    This function takes a file location as a string as an argument. The argument is passed on to the data_list\n",
    "    function to obtain lst_genes, lst_ge_C, and lst_ge_AD. Using the compute_row_average function, the mean\n",
    "    RPMK values are obtained. The log base 2 of the ratio of disease RPMK to control RPMK is determined. Genes\n",
    "    with a an absolute Log base 2 ratio greater than or equal to 1 are added to the dictionary, gene_dict, along\n",
    "    with their values. Negative values indicate genes in which expression is two-fold or greater in the control\n",
    "    compared to the disease state. Positive values indicate genes in which expression is two-fold or greater in\n",
    "    the disease state compared to the control.\n",
    "    \n",
    "    \n",
    "    \n",
    "    Parameters: \n",
    "    filename (string): The argument of this function is a an RNAseq data (.txt) file location as a string.\n",
    "\n",
    "\n",
    "    Returns: \n",
    "    gene_dict (dictionary): This output of this function is a dictionary of genes with a 2-fold or more increase\n",
    "                            of gene expression in disease states compared to control.\n",
    "\n",
    "    \"\"\"\n",
    "    lst_genes, lst_ge_C, lst_ge_AD = data_list(file_name)\n",
    "    ge_mean_C = compute_row_average(lst_ge_C)\n",
    "    ge_mean_AD = compute_row_average(lst_ge_AD)\n",
    "    ge_mean_C += 1\n",
    "    ge_mean_AD +=1\n",
    "    RPMKratio= np.divide(ge_mean_AD,ge_mean_C)\n",
    "    RPMKratio_log=[]\n",
    "    gene_dict=dict()\n",
    "    for element in RPMKratio:\n",
    "        b2log=math.log(element,2)\n",
    "        #b2log=abs(b2log)\n",
    "        RPMKratio_log.append(b2log)\n",
    "    for element in RPMKratio_log:\n",
    "        if abs(element) >= 1:\n",
    "            position = RPMKratio_log.index(element)\n",
    "            gene_dict[lst_genes[position]]= element     \n",
    "    return gene_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "----"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'B2M': -1.0831605463018765,\n",
       " 'CD44': 1.1560769875383454,\n",
       " 'CHI3L1': -1.8047781530898872,\n",
       " 'CHI3L2': -2.707243055240693,\n",
       " 'DSP': -1.1441831930663147,\n",
       " 'IFI6': -1.5164312668299031,\n",
       " 'GBP1': -1.5800630930378448,\n",
       " 'GBP2': -1.1175208782755517,\n",
       " 'GBP3': -1.0950696178929018,\n",
       " 'IFIT3': -1.4148330181938937,\n",
       " 'IGF2': -1.0472490118066526,\n",
       " 'INDO': -1.6683566883503855,\n",
       " 'CXCL10': -1.885573593462525,\n",
       " 'ITIH4': -1.1331664356985078,\n",
       " 'MGP': -1.6161683252017167,\n",
       " 'CXCL9': -3.856847738539151,\n",
       " 'OAS1': -1.6334446059020005,\n",
       " 'OAS3': -1.0420785413245308,\n",
       " 'RARRES3': -1.2606496379935344,\n",
       " 'S100A8': -1.624071217798039,\n",
       " 'S100A9': -1.6345458580287295,\n",
       " 'S100A12': -1.37429296546868,\n",
       " 'SAA1': -1.2798556154280047,\n",
       " 'CCL2': 1.304777297411683,\n",
       " 'SLPI': -2.0716382167069907,\n",
       " 'TAP1': -1.0301873977834999,\n",
       " 'TTR': -1.05871241958657,\n",
       " 'UTY': -1.1942877312192577,\n",
       " 'JARID1D': -2.0197058572193796,\n",
       " 'USP9Y': -1.0501526497324956,\n",
       " 'DDX3Y': -1.393925707894159,\n",
       " 'ISG15': -2.6079770382911507,\n",
       " 'IFI44L': -1.0851365892348082,\n",
       " 'BTN3A2': -1.071294141819273,\n",
       " 'RSAD2': -1.4368990142316904,\n",
       " 'GNG8': -1.0042846587266263,\n",
       " 'GBP5': -1.1109237554385751,\n",
       " 'BATF2': -1.0336541889695072,\n",
       " 'H19': -1.4510652482519526,\n",
       " 'GOLGA8B': -1.0797355893649876,\n",
       " 'FAM153C': -1.0257828707768832,\n",
       " 'SNORA74B': -1.2565581890295294}"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Do not change this cell\n",
    "gene_dict = gene_dictionary(\"GSE53697_RNAseq_AD.txt\")\n",
    "gene_dict"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
