{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Homework 6 - BIOINF 575\n",
    "\n",
    "Goals:\n",
    "* Reading from a file with pandas\n",
    "* Creating and using pd.DataFrame \n",
    "* Saving results to a file\n",
    "\n",
    "This homework even though it has two problems should be very short in terms of lines of code to write.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Problem 1. \n",
    "\n",
    "The data file (`GSE53697_RNAseq_AD.txt`) is a tab-separated file that contains raw read counts and gene expression values in the form of RPKM values from RNA sequencing of 8 control and 9 Alzheimer's disease (AD) samples (human).\n",
    "\n",
    "The first line in the file is the header that describes the table of values:\n",
    "\n",
    "    GeneID - Gene ID\n",
    "    GeneSymbol - Gene Symbol\n",
    "    8 columns: C1-8_raw \n",
    "        raw read count values for the 8 control samples\n",
    "    9 columns: A9-17_raw \n",
    "        raw read count values for the 9 Alzheimer's disease samples\n",
    "    8 columns: C1-8_rpkm \n",
    "        RPKM values for the 8 control samples\n",
    "    9 columns: A9-17_rpkm \n",
    "        RPKM values for the 9 Alzheimer's disease (AD) samples\n",
    "\n",
    "Reads Per Kilobase of transcript per Million mapped reads (RPKM) is a normalized unit of transcript/gene expression. This is used to normalize for sequencing depth and gene length. Different genes might have more reads just due to the fact that they have more bases that the reads can align to and that needs to be accounted for. Also, if one sample (s1) has more reads than another sample (s2) then the overall expression of all genes in s1 would be higher than in s2. Hence, the two samples cannot be compared if a normalization for sequencing depth is not performed.\n",
    "\n",
    "Given this file, select differentially expressed genes between Alzheimer's disease and control. There are many ways to compute differential expression. Here, to go from expression to differential expression, for each gene you have to compute the ratio between two averages: (1) the average expression of the gene in disease and (2) the average expression of the gene in control.\n",
    "\n",
    "Since these averages might be 0 the division by zero will result in infinite (na.inf, which can be checked with the funtion:  na.isfinite). Ignore the infinite values. Select the differentially expressed genes and return them as a pd.DataFrame.\n",
    "\n",
    "#### Assignment Requirement:\n",
    "\n",
    "* Create a function `compute_DEgenes` that prcesses a file with the file name given as a parameter and computes the a data frame with the differentially expressed genes (genes with a log2 fold change > 1, where the fold change is the ratio between the average disease RPKM values and the average of the control RPKM values). The data frame contains only on column which is the values of the log2ratios and the row names that are the gene symbols.\n",
    "* Test the function and write the results to a file \"DEgenes.csv\".\n",
    "\n",
    "As long as your solution makes use of `pd.DataFrame` and returns the required output it is an acceptable solution.\n",
    "\n",
    "One way to do this is:\n",
    "* Read the file content into a `pd.DataFrame`. \n",
    "* Use subsetting to create a new data frame with the rpkm control values and compute the row means. \n",
    "* Use subsetting to create a new data frame with the rpkm AD values and compute the row means. \n",
    "* Create a new data frame with the log2 ratios and the gene symbols as row names (gene symbols are available in the initial data frame where the data was read from the file).\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Problem 2. \n",
    "\n",
    "The data you will be working with (`clinvar_20190923_short.vcf`) contains several allele frequencies from different databases. The one to look for in this assignment is from ExAC database.\n",
    "\n",
    "The beginning of every VCF file contains various sets of information:\n",
    "* Meta-information (details about the experiment or configuration) lines start with **`##`**\n",
    "    * These lines are helpful in understanding specialized keys found in the `INFO` column. \n",
    "* Header lines (column names) start with **`#`**\n",
    "\n",
    "From there on, each line is made up of tab (`\\t`) separated values that make up eight (8) columns. Those columns are:\n",
    "    1. CHROM (chromosome)\n",
    "    2. POS (base pair position of the variant)\n",
    "    3. ID (identifier if applicable; `.` if not applicable/missing)\n",
    "    4. REF (reference base)\n",
    "    5. ALT (alternate base(s): comma (`,`) separated if applicable)\n",
    "    6. QUAL (Phred-scaled quality score; `.` if not applicable/missing)\n",
    "    7. FILTER (filter status; `.` if not applicable/missing)\n",
    "    8. INFO (any additional information about the variant)\n",
    "    * Semi-colon (`;`) separated key-value pairs\n",
    "    * Key-value pairs are equal sign (`=`) separated \n",
    "        (key on the left, value on the right)\n",
    "    * If a key has multiple values, the values are comma (`|`) separated\n",
    "\n",
    "There are some additional details to consider for this assignment. You will be expected to consider two (2) special types of keys:\n",
    "* The `AF_EXAC` key that describes the allele frequencies from the ExAC database\n",
    "    > `##INFO=<ID=AF_EXAC,Number=1,Type=Float,Description=\"allele frequencies from ExAC\">`\n",
    "    * The data included are `float`ing point numbers\n",
    "* The `CLNDN` key that gives all the names the given variant is associated with\n",
    "    > `##INFO=<ID=CLNDN,Number=.,Type=String,Description=\"ClinVar's preferred disease name for the concept specified by disease identifiers in CLNDISDB\">`\n",
    "    * The data are`str`ings. **However**, if there are multiple diseases associated with a given variant, the diseases are pipe (`|`) separated.\n",
    "\n",
    "#### Assignment Requirement\n",
    "* Write a funtion `select_variants` that takes the file name as a parameter and returns a data frame with rare variants (AF_EXAC < 0.0001) that are associated with at least 2 diseases have `\"C\"` as REF and `\"T\"` as a ALT.\n",
    "* Test the function and write the results to a file \"SELvariants.csv\".\n",
    "\n",
    "As long as your solution makes use of `pd.DataFrame` and returns the required output it is an acceptable solution.\n",
    "\n",
    "One way to do this is:\n",
    "* Read the table from the file into a pd.DataFrame. The `pd.read_csv()` funtion has  `comment` parameter that can be used to skip the lies that start with `\"#\"`. However, columns labels have to be added by using the `columns` attribute if that is done.\n",
    "* Similar to homework 3, compute a list of diseaseses for rare variants (variants with AF_EXAC < 0.0001), the funtion will return an empty list if the variant is not rare or does not have any disease associated. The `not_specified` and `not_provided` should not appear in these lists. Use the details provided in homework 3. You can do this part of the homework by adapting the parse_line function you wrote for homework 3 to a parse_INFO function that does not parse a line but the value in the INFO column, since that is now a column in the data frame. \n",
    "* The apply function of a pd.Series can be used to apply the funtion parse_INFO to each element of the INFO column of the data frame to compute the disease list for each variant and the len funtion can be applied to the resulting pd.Series to compute the length of the disease list for each variant. \n",
    "* The result can be stored as a new column in the initial data frame. The new column contains the length of the associated disease list.\n",
    "* Subset the updated data frame using conditional subsetting. Conditions on the `REF` in `ALT` and the newly added column (that contains the associated disease list length) would compute a data frame with rare variants (AF_EXAC < 0.0001) that are associated with at least 2 diseases, have `\"C\"` as REF and `\"T\"` as a ALT. \n",
    "\n",
    "    \n",
    "    \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "---\n",
    "## Academic Honor Code\n",
    "In accordance with Rackham's Academic Misconduct Policy; upon submission of your assignment, you (the student) are indicating acceptance of the following statement:\n",
    "\n",
    "> “I pledge that this submission is solely my own work.”\n",
    "\n",
    "As such, the instructors reserve the right to process any and all source code therein contained within the submitted notebooks with source code plagiarism detection software.\n",
    "\n",
    "Any violations of the this agreement will result in swift, sure, and significant punishment."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_DEgenes(file_name):\n",
    "    '''\n",
    "    This function will parse for genes of which have a 2-fold expression change\n",
    "    (positive or negative log) between a disease and control state.\n",
    "    \n",
    "\n",
    "    This function takes a file (.txt) location as a string to open and read in the data to a DataFrame from said file.\n",
    "    A value of 1 is added to both the control and disease RPKM values(to avoid ZeroDivisionError when calculating\n",
    "    the ratio). The ratio between the disease RPKM to control RPKM values is computed and the log base 2 of the\n",
    "    ratio is added to the DataFrame. Genes with an absolute log base 2 ratio greater than 1 are differentiated for\n",
    "    and saved to a new DataFrame. The new DataFrame is written to a .csv file, 'DEgenes.csv'\n",
    "\n",
    "    Parameters: \n",
    "    filename (string): The argument of this function is (.txt) file location as a string.\n",
    "\n",
    "\n",
    "    Returns: \n",
    "    genes_de_df (DataFrame): This output is a DataFrame of the genes with an absolute log base 2 ratio of\n",
    "                             disease RPKM to control RPKM values expression greater than 1\n",
    "    DEgenes.csv (file.csv): This output is a .csv file of the genes_de_df DataFrame\n",
    "    \n",
    "    '''\n",
    "    data=pd.read_csv('GSE53697_RNAseq_AD.txt', header=0, sep='\\t',index_col=1)\n",
    "    crpmk=data[['C1_rpkm', 'C2_rpkm','C3_rpkm','C4_rpkm','C5_rpkm','C6_rpkm','C7_rpkm','C8_rpkm']]\n",
    "    for row in crpmk:\n",
    "        crpmk_mean= crpmk.mean(axis=1)\n",
    "        crpmk_mean_PlusValue1 = crpmk_mean +1\n",
    "    crpmk_mean_list=crpmk_mean.tolist()\n",
    "    crpmk_mean_PlusValue1_list=crpmk_mean_PlusValue1.tolist()\n",
    "    data['crpmk_mean']=crpmk_mean_list\n",
    "    data['crpmk_mean_PlusValue1']=crpmk_mean_PlusValue1_list\n",
    "    Arpmk=data[['A9_rpkm', 'A10_rpkm','A11_rpkm','A12_rpkm','A13_rpkm','A14_rpkm','A15_rpkm','A16_rpkm','A17_rpkm']]\n",
    "    for row in Arpmk:\n",
    "        Arpmk_mean= Arpmk.mean(axis=1)\n",
    "        Arpmk_mean_PlusValue1 = Arpmk_mean +1   \n",
    "    Arpmk_mean_list=Arpmk_mean.tolist()\n",
    "    Arpmk_mean_PlusValue1_list= Arpmk_mean_PlusValue1.tolist()\n",
    "    data['Arpmk_mean']=Arpmk_mean_list\n",
    "    data['Arpmk_mean_PlusValue1']= Arpmk_mean_PlusValue1_list\n",
    "    Arpmk_mean_array= np.array(Arpmk_mean_PlusValue1_list)\n",
    "    crpmk_mean_array=np.array(crpmk_mean_PlusValue1_list)\n",
    "    log2ratios=np.log2(np.divide(Arpmk_mean_array,crpmk_mean_array))\n",
    "    data['log2ratios']=log2ratios\n",
    "    data1=data.copy()\n",
    "    data1=data1.loc[abs(data1['log2ratios']) >1]\n",
    "    dif_expressions_data=data1['log2ratios'] #when index for a specific column, it turns it into a series, not a dataframe\n",
    "    genes_de_df=dif_expressions_data.to_frame()\n",
    "    genes_de_df.to_csv('DEgenes.csv')\n",
    "    return genes_de_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>log2ratios</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GeneSymbol</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>B2M</td>\n",
       "      <td>-1.083161</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CD44</td>\n",
       "      <td>1.156077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CHI3L1</td>\n",
       "      <td>-1.804778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CHI3L2</td>\n",
       "      <td>-2.707243</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>DSP</td>\n",
       "      <td>-1.144183</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>IFI6</td>\n",
       "      <td>-1.516431</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>GBP1</td>\n",
       "      <td>-1.580063</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>GBP2</td>\n",
       "      <td>-1.117521</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>GBP3</td>\n",
       "      <td>-1.095070</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>IFIT3</td>\n",
       "      <td>-1.414833</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>IGF2</td>\n",
       "      <td>-1.047249</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>INDO</td>\n",
       "      <td>-1.668357</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CXCL10</td>\n",
       "      <td>-1.885574</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ITIH4</td>\n",
       "      <td>-1.133166</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>MGP</td>\n",
       "      <td>-1.616168</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CXCL9</td>\n",
       "      <td>-3.856848</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>OAS1</td>\n",
       "      <td>-1.633445</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>OAS3</td>\n",
       "      <td>-1.042079</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>RARRES3</td>\n",
       "      <td>-1.260650</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>S100A8</td>\n",
       "      <td>-1.624071</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>S100A9</td>\n",
       "      <td>-1.634546</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>S100A12</td>\n",
       "      <td>-1.374293</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>SAA1</td>\n",
       "      <td>-1.279856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CCL2</td>\n",
       "      <td>1.304777</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>SLPI</td>\n",
       "      <td>-2.071638</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>TAP1</td>\n",
       "      <td>-1.030187</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>TTR</td>\n",
       "      <td>-1.058712</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>UTY</td>\n",
       "      <td>-1.194288</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>JARID1D</td>\n",
       "      <td>-2.019706</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>USP9Y</td>\n",
       "      <td>-1.050153</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>DDX3Y</td>\n",
       "      <td>-1.393926</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ISG15</td>\n",
       "      <td>-2.607977</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>IFI44L</td>\n",
       "      <td>-1.085137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>BTN3A2</td>\n",
       "      <td>-1.071294</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>RSAD2</td>\n",
       "      <td>-1.436899</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>GNG8</td>\n",
       "      <td>-1.004285</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>GBP5</td>\n",
       "      <td>-1.110924</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>BATF2</td>\n",
       "      <td>-1.033654</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>H19</td>\n",
       "      <td>-1.451065</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>GOLGA8B</td>\n",
       "      <td>-1.079736</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>FAM153C</td>\n",
       "      <td>-1.025783</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>SNORA74B</td>\n",
       "      <td>-1.256558</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            log2ratios\n",
       "GeneSymbol            \n",
       "B2M          -1.083161\n",
       "CD44          1.156077\n",
       "CHI3L1       -1.804778\n",
       "CHI3L2       -2.707243\n",
       "DSP          -1.144183\n",
       "IFI6         -1.516431\n",
       "GBP1         -1.580063\n",
       "GBP2         -1.117521\n",
       "GBP3         -1.095070\n",
       "IFIT3        -1.414833\n",
       "IGF2         -1.047249\n",
       "INDO         -1.668357\n",
       "CXCL10       -1.885574\n",
       "ITIH4        -1.133166\n",
       "MGP          -1.616168\n",
       "CXCL9        -3.856848\n",
       "OAS1         -1.633445\n",
       "OAS3         -1.042079\n",
       "RARRES3      -1.260650\n",
       "S100A8       -1.624071\n",
       "S100A9       -1.634546\n",
       "S100A12      -1.374293\n",
       "SAA1         -1.279856\n",
       "CCL2          1.304777\n",
       "SLPI         -2.071638\n",
       "TAP1         -1.030187\n",
       "TTR          -1.058712\n",
       "UTY          -1.194288\n",
       "JARID1D      -2.019706\n",
       "USP9Y        -1.050153\n",
       "DDX3Y        -1.393926\n",
       "ISG15        -2.607977\n",
       "IFI44L       -1.085137\n",
       "BTN3A2       -1.071294\n",
       "RSAD2        -1.436899\n",
       "GNG8         -1.004285\n",
       "GBP5         -1.110924\n",
       "BATF2        -1.033654\n",
       "H19          -1.451065\n",
       "GOLGA8B      -1.079736\n",
       "FAM153C      -1.025783\n",
       "SNORA74B     -1.256558"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# test the function compute_DEgenes with the given file 'GSE53697_RNAseq_AD.txt'\n",
    "\n",
    "# save result to file \"DEgenes.csv\"\n",
    "\n",
    "compute_DEgenes('GSE53697_RNAseq_AD.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_info(INFO):\n",
    "    '''\n",
    "    This function will create a list of known diseases associated to rare gene variants, as determined by an \n",
    "    AF_EXAC value < 0.0001. If the variant is not rare, or there is no known associated disease, then an empty\n",
    "    list is returned\n",
    "    \n",
    "\n",
    "    This function takes a pd.series containing the INFO column of a .vcf file and parses the input for rare\n",
    "    gene variants, as determined by an AF_EXAC value < 0.0001. The associated diseases of the variant are then\n",
    "    returned in a list. If a gene variant is not rare, or has no known associated diseases, an empty list is\n",
    "    returned\n",
    "\n",
    "    Parameters: \n",
    "    INFO (pd.series): The argument of this function is a pd.series containing the INFO column of a .vcf file\n",
    "\n",
    "\n",
    "    Returns: \n",
    "    CLNDN_disease_clean (list): This output is a list of the associated diseases to a rare gene variant. If the\n",
    "                                variant is not rare, or there are no known associated diseases, the list is \n",
    "                                created and is returned empty\n",
    "    \n",
    "    '''\n",
    "    INFO=INFO.split()\n",
    "    for element in INFO:\n",
    "        if not'AF_EXAC' in element:\n",
    "            CLNDN_disease_clean=[]\n",
    "        else:\n",
    "            datapoint=element.split(';')\n",
    "            for elementD in datapoint:\n",
    "                if elementD.startswith(\"AF_EXAC\") ==True: #index to AF_EXAC column and compare values\n",
    "                    AF_EXAC=elementD.split(\"=\")\n",
    "                    AF_EXACnum=float(AF_EXAC[1])\n",
    "                    if AF_EXACnum > 0.0001:\n",
    "                        CLNDN_disease_clean=[]\n",
    "                    elif AF_EXACnum == 0.0001 or AF_EXACnum< 0.0001:\n",
    "                        for elementC in datapoint:\n",
    "                            if elementC.startswith(\"CLNDN\") == True: #Index to diseases to create list\n",
    "                                CLNDN_list = elementC.split('=')\n",
    "                                CLNDN_disease=CLNDN_list[1]\n",
    "                                CLNDN_disease1=CLNDN_disease.split('|')\n",
    "                                CLNDN_disease_clean=[]\n",
    "                                for elementCD in CLNDN_disease1:\n",
    "                                    if elementCD != 'not_provided' and elementCD !='not_specified':\n",
    "                                        CLNDN_disease_clean.append(elementCD)\n",
    "    return(CLNDN_disease_clean)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def select_variants(filename):\n",
    "    '''\n",
    "    This function will parse for rare gene variants determined by an AF_EXAC value < 0.0001 and variants with \n",
    "    more than 1 known associated disease within that pool.\n",
    "    \n",
    "\n",
    "    This function takes a .vcf file location as a string to open and read in the data to a DataFrame from said file.\n",
    "    A pd.series is created from the INFO column of the DataFrame and the function, parse_info, is applied to the series.\n",
    "    A new column, ASSOCIATED_DISEASE_NUMBER, is added to the initial DataFrame(using the length of the pd.series after\n",
    "    parse_info is applied). A smaller DataFrame containing only the rare variants with more than one known associated\n",
    "    disease is created(ref_variants). This DataFrame is then selected for only the gene variants with a REF value of 'C'\n",
    "    and an ALT value of 'T' and saved into a new DataFrame, variants_selected_df, which is saved to the file \n",
    "    'SELvariants.csv'\n",
    "\n",
    "    Parameters: \n",
    "    filename (string): The argument of this function is a file (.vcf) location as a string.\n",
    "\n",
    "\n",
    "    Returns: \n",
    "    variants_selected_df (DataFrame): This output is a DataFrame of the rare gene variants(as determined by an \n",
    "                                      AF_EXAC value < 0.0001) with more than one known associated disease,\n",
    "                                      a REF value of 'C', and an ALT value of 'T'\n",
    "    DEgenes.csv (file.csv): This output is a .csv file of the variants_selected_df DataFrame\n",
    "    \n",
    "    '''\n",
    "    clinvar_initial=pd.read_csv(filename, sep='\\t', comment='#', names=['CHROM','POS','ID','REF','ALT','QUAL','FILTER','INFO'])\n",
    "    clinvar_INFO=clinvar_initial['INFO']\n",
    "    clinvar_INFO_applied=clinvar_INFO.apply(parse_info)\n",
    "    clinvar_initial['ASSOCIATED_DISEASE_NUMBER']=clinvar_INFO_applied.apply(len)\n",
    "    clinvar_rare_disease=clinvar_initial.loc[clinvar_initial['ASSOCIATED_DISEASE_NUMBER']>1]\n",
    "    ref_variants=clinvar_rare_disease.loc[clinvar_rare_disease['REF'] == 'C']\n",
    "    variants_selected_df=ref_variants.loc[ref_variants['ALT'] == 'T']\n",
    "    variants_selected_df.to_csv('SELvariants.csv')\n",
    "    \n",
    "    \n",
    "    return variants_selected_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>INFO</th>\n",
       "      <th>ASSOCIATED_DISEASE_NUMBER</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>632</td>\n",
       "      <td>1</td>\n",
       "      <td>1806513</td>\n",
       "      <td>224713</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>AF_EXAC=0.00001;ALLELEID=226500;CLNDISDB=.|.|H...</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>795</td>\n",
       "      <td>1</td>\n",
       "      <td>2229492</td>\n",
       "      <td>381375</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>AF_ESP=0.00008;AF_EXAC=0.00005;ALLELEID=364674...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>818</td>\n",
       "      <td>1</td>\n",
       "      <td>2229730</td>\n",
       "      <td>264199</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>AF_EXAC=0.00000;ALLELEID=257945;CLNDISDB=MedGe...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>823</td>\n",
       "      <td>1</td>\n",
       "      <td>2302983</td>\n",
       "      <td>520164</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>AF_EXAC=0.00002;ALLELEID=509149;CLNDISDB=MedGe...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>920</td>\n",
       "      <td>1</td>\n",
       "      <td>2304437</td>\n",
       "      <td>213697</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>AF_EXAC=0.00000;ALLELEID=209481;CLNDISDB=MedGe...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>922</td>\n",
       "      <td>1</td>\n",
       "      <td>2304450</td>\n",
       "      <td>213678</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>AF_ESP=0.00008;AF_EXAC=0.00009;ALLELEID=209482...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>950</td>\n",
       "      <td>1</td>\n",
       "      <td>2306158</td>\n",
       "      <td>520169</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>AF_ESP=0.00008;AF_EXAC=0.00002;ALLELEID=509157...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     CHROM      POS      ID REF ALT QUAL FILTER  \\\n",
       "632      1  1806513  224713   C   T    .      .   \n",
       "795      1  2229492  381375   C   T    .      .   \n",
       "818      1  2229730  264199   C   T    .      .   \n",
       "823      1  2302983  520164   C   T    .      .   \n",
       "920      1  2304437  213697   C   T    .      .   \n",
       "922      1  2304450  213678   C   T    .      .   \n",
       "950      1  2306158  520169   C   T    .      .   \n",
       "\n",
       "                                                  INFO  \\\n",
       "632  AF_EXAC=0.00001;ALLELEID=226500;CLNDISDB=.|.|H...   \n",
       "795  AF_ESP=0.00008;AF_EXAC=0.00005;ALLELEID=364674...   \n",
       "818  AF_EXAC=0.00000;ALLELEID=257945;CLNDISDB=MedGe...   \n",
       "823  AF_EXAC=0.00002;ALLELEID=509149;CLNDISDB=MedGe...   \n",
       "920  AF_EXAC=0.00000;ALLELEID=209481;CLNDISDB=MedGe...   \n",
       "922  AF_ESP=0.00008;AF_EXAC=0.00009;ALLELEID=209482...   \n",
       "950  AF_ESP=0.00008;AF_EXAC=0.00002;ALLELEID=509157...   \n",
       "\n",
       "     ASSOCIATED_DISEASE_NUMBER  \n",
       "632                          6  \n",
       "795                          2  \n",
       "818                          2  \n",
       "823                          2  \n",
       "920                          2  \n",
       "922                          2  \n",
       "950                          2  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# test the function select_variants with the given file 'clinvar_20190923_short.vcf'\n",
    "\n",
    "# save result to file \"SELvariants.csv\"\n",
    "\n",
    "select_variants('clinvar_20190923_short.vcf')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "-----------"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>log2ratios</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GeneSymbol</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>B2M</td>\n",
       "      <td>-1.083161</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CD44</td>\n",
       "      <td>1.156077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CHI3L1</td>\n",
       "      <td>-1.804778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CHI3L2</td>\n",
       "      <td>-2.707243</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>DSP</td>\n",
       "      <td>-1.144183</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>IFI6</td>\n",
       "      <td>-1.516431</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>GBP1</td>\n",
       "      <td>-1.580063</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>GBP2</td>\n",
       "      <td>-1.117521</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>GBP3</td>\n",
       "      <td>-1.095070</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>IFIT3</td>\n",
       "      <td>-1.414833</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>IGF2</td>\n",
       "      <td>-1.047249</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>INDO</td>\n",
       "      <td>-1.668357</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CXCL10</td>\n",
       "      <td>-1.885574</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ITIH4</td>\n",
       "      <td>-1.133166</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>MGP</td>\n",
       "      <td>-1.616168</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CXCL9</td>\n",
       "      <td>-3.856848</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>OAS1</td>\n",
       "      <td>-1.633445</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>OAS3</td>\n",
       "      <td>-1.042079</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>RARRES3</td>\n",
       "      <td>-1.260650</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>S100A8</td>\n",
       "      <td>-1.624071</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>S100A9</td>\n",
       "      <td>-1.634546</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>S100A12</td>\n",
       "      <td>-1.374293</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>SAA1</td>\n",
       "      <td>-1.279856</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>CCL2</td>\n",
       "      <td>1.304777</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>SLPI</td>\n",
       "      <td>-2.071638</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>TAP1</td>\n",
       "      <td>-1.030187</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>TTR</td>\n",
       "      <td>-1.058712</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>UTY</td>\n",
       "      <td>-1.194288</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>JARID1D</td>\n",
       "      <td>-2.019706</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>USP9Y</td>\n",
       "      <td>-1.050153</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>DDX3Y</td>\n",
       "      <td>-1.393926</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>ISG15</td>\n",
       "      <td>-2.607977</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>IFI44L</td>\n",
       "      <td>-1.085137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>BTN3A2</td>\n",
       "      <td>-1.071294</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>RSAD2</td>\n",
       "      <td>-1.436899</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>GNG8</td>\n",
       "      <td>-1.004285</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>GBP5</td>\n",
       "      <td>-1.110924</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>BATF2</td>\n",
       "      <td>-1.033654</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>H19</td>\n",
       "      <td>-1.451065</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>GOLGA8B</td>\n",
       "      <td>-1.079736</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>FAM153C</td>\n",
       "      <td>-1.025783</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>SNORA74B</td>\n",
       "      <td>-1.256558</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            log2ratios\n",
       "GeneSymbol            \n",
       "B2M          -1.083161\n",
       "CD44          1.156077\n",
       "CHI3L1       -1.804778\n",
       "CHI3L2       -2.707243\n",
       "DSP          -1.144183\n",
       "IFI6         -1.516431\n",
       "GBP1         -1.580063\n",
       "GBP2         -1.117521\n",
       "GBP3         -1.095070\n",
       "IFIT3        -1.414833\n",
       "IGF2         -1.047249\n",
       "INDO         -1.668357\n",
       "CXCL10       -1.885574\n",
       "ITIH4        -1.133166\n",
       "MGP          -1.616168\n",
       "CXCL9        -3.856848\n",
       "OAS1         -1.633445\n",
       "OAS3         -1.042079\n",
       "RARRES3      -1.260650\n",
       "S100A8       -1.624071\n",
       "S100A9       -1.634546\n",
       "S100A12      -1.374293\n",
       "SAA1         -1.279856\n",
       "CCL2          1.304777\n",
       "SLPI         -2.071638\n",
       "TAP1         -1.030187\n",
       "TTR          -1.058712\n",
       "UTY          -1.194288\n",
       "JARID1D      -2.019706\n",
       "USP9Y        -1.050153\n",
       "DDX3Y        -1.393926\n",
       "ISG15        -2.607977\n",
       "IFI44L       -1.085137\n",
       "BTN3A2       -1.071294\n",
       "RSAD2        -1.436899\n",
       "GNG8         -1.004285\n",
       "GBP5         -1.110924\n",
       "BATF2        -1.033654\n",
       "H19          -1.451065\n",
       "GOLGA8B      -1.079736\n",
       "FAM153C      -1.025783\n",
       "SNORA74B     -1.256558"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Do not change this cell\n",
    "genes_df = pd.read_csv(\"DEgenes.csv\", index_col=0)\n",
    "genes_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>INFO</th>\n",
       "      <th>ASSOCIATED_DISEASE_NUMBER</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>632</td>\n",
       "      <td>1</td>\n",
       "      <td>1806513</td>\n",
       "      <td>224713</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>AF_EXAC=0.00001;ALLELEID=226500;CLNDISDB=.|.|H...</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>795</td>\n",
       "      <td>1</td>\n",
       "      <td>2229492</td>\n",
       "      <td>381375</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>AF_ESP=0.00008;AF_EXAC=0.00005;ALLELEID=364674...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>818</td>\n",
       "      <td>1</td>\n",
       "      <td>2229730</td>\n",
       "      <td>264199</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>AF_EXAC=0.00000;ALLELEID=257945;CLNDISDB=MedGe...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>823</td>\n",
       "      <td>1</td>\n",
       "      <td>2302983</td>\n",
       "      <td>520164</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>AF_EXAC=0.00002;ALLELEID=509149;CLNDISDB=MedGe...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>920</td>\n",
       "      <td>1</td>\n",
       "      <td>2304437</td>\n",
       "      <td>213697</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>AF_EXAC=0.00000;ALLELEID=209481;CLNDISDB=MedGe...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>922</td>\n",
       "      <td>1</td>\n",
       "      <td>2304450</td>\n",
       "      <td>213678</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>AF_ESP=0.00008;AF_EXAC=0.00009;ALLELEID=209482...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>950</td>\n",
       "      <td>1</td>\n",
       "      <td>2306158</td>\n",
       "      <td>520169</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>AF_ESP=0.00008;AF_EXAC=0.00002;ALLELEID=509157...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0  CHROM      POS      ID REF ALT QUAL FILTER  \\\n",
       "0         632      1  1806513  224713   C   T    .      .   \n",
       "1         795      1  2229492  381375   C   T    .      .   \n",
       "2         818      1  2229730  264199   C   T    .      .   \n",
       "3         823      1  2302983  520164   C   T    .      .   \n",
       "4         920      1  2304437  213697   C   T    .      .   \n",
       "5         922      1  2304450  213678   C   T    .      .   \n",
       "6         950      1  2306158  520169   C   T    .      .   \n",
       "\n",
       "                                                INFO  \\\n",
       "0  AF_EXAC=0.00001;ALLELEID=226500;CLNDISDB=.|.|H...   \n",
       "1  AF_ESP=0.00008;AF_EXAC=0.00005;ALLELEID=364674...   \n",
       "2  AF_EXAC=0.00000;ALLELEID=257945;CLNDISDB=MedGe...   \n",
       "3  AF_EXAC=0.00002;ALLELEID=509149;CLNDISDB=MedGe...   \n",
       "4  AF_EXAC=0.00000;ALLELEID=209481;CLNDISDB=MedGe...   \n",
       "5  AF_ESP=0.00008;AF_EXAC=0.00009;ALLELEID=209482...   \n",
       "6  AF_ESP=0.00008;AF_EXAC=0.00002;ALLELEID=509157...   \n",
       "\n",
       "   ASSOCIATED_DISEASE_NUMBER  \n",
       "0                          6  \n",
       "1                          2  \n",
       "2                          2  \n",
       "3                          2  \n",
       "4                          2  \n",
       "5                          2  \n",
       "6                          2  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Do not change this cell\n",
    "variants_df = pd.read_csv(\"SELvariants.csv\")\n",
    "variants_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
